toxics/timeout_test.go

```diff
@@ -0,0 +1,117 @@
+package toxics_test
+
+import (
+	"io"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/Shopify/toxiproxy/toxics"
+)
+
+func TestTimeoutToxicDoesNotCauseHang(t *testing.T) {
+	ln, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatal("Failed to create TCP server", err)
+	}
+	defer ln.Close()
+
+	proxy := NewTestProxy("test", ln.Addr().String())
+	proxy.Start()
+	defer proxy.Stop()
+
+	proxy.Toxics.AddToxicJson(ToxicToJson(t, "might_block", "latency", "upstream", &toxics.LatencyToxic{Latency: 10}))
+	proxy.Toxics.AddToxicJson(ToxicToJson(t, "to_delete", "timeout", "upstream", &toxics.TimeoutToxic{Timeout: 0}))
+
+	serverConnRecv := make(chan net.Conn)
+	go func() {
+		conn, err := ln.Accept()
+		if err != nil {
+			t.Error("Unable to accept TCP connection", err)
+		}
+		serverConnRecv <- conn
+	}()
+
+	conn, err := net.Dial("tcp", proxy.Listen)
+	if err != nil {
+		t.Fatal("Unable to dial TCP server", err)
+	}
+	defer conn.Close()
+
+	_ = <-serverConnRecv
+
+	_, err = conn.Write([]byte("hello"))
+	if err != nil {
+		t.Fatal("Unable to write to proxy", err)
+	}
+
+	time.Sleep(1 * time.Second) // Shitty sync waiting for latency toxi to get data.
```

Comment:
This frustrates me.

Our bug only occurs when the `latency` toxic is blocked attempting to send data to the `timeout` toxic. Without the sleep, the test removes the `latency` toxic before it receives data which means the bug does not occur because the `latency` toxic is not blocking.

I can't think of a non racy way to wait for the `latency` toxic to receive input data. We can't wait for the upstream to receive the data because the `timeout` toxic is blocking data from reaching the upstream.